pkg/controller/operators/catalog/operator.go

```diff
@@ -1125,10 +1145,37 @@ func (o *Operator) validateExistingCRs(gvr schema.GroupVersionResource, newCRD *
 			return fmt.Errorf("error validating custom resource against new schema %#v: %s", newCRD.Spec.Validation, err)
 		}
 	}
-
 	return nil
 }
 
+// Attempt to remove stored versions that have been deprecated before allowing
+// those versions to be removed from the new CRD.
+// The function may not always succeed as storedVersions requires at least one
+// version. If there is only stored version, it won't be removed until a new
+// stored version is added.
+func removeDeprecatedStoredVersions(oldCRD *v1beta1ext.CustomResourceDefinition, newCRD *v1beta1ext.CustomResourceDefinition) *v1beta1ext.CustomResourceDefinition {
```

Comment:
this is taking a reference to `oldCRD` and modifying it, and also returning that same modified one. I think we should pick one or the other - either pass a ref and modify it (and don't return it) or make a copy and return the copy